File: memory-domain/src/main/java/com/memory/domain/memory/repository/MemoryRepositoryCustom.java
package com.memory.domain.memory.repository;

import com.memory.domain.memory.Memory;
import com.memory.domain.member.Member;
import com.memory.domain.memory.MemoryType;

import java.util.List;
import java.util.Optional;

public interface MemoryRepositoryCustom {
    List<Memory> findByMemberAndMemoryType(Member member, MemoryType memoryType, int size);
    List<Memory> findByMemberAndMemoryType(Member member, MemoryType memoryType, Long lastMemoryId, int size);
    Optional<Memory> findMemoryByIdAndMemberId(Long memoryId, Long memberId);
    List<Memory> findByMemoryType(MemoryType memoryType, int size);
    List<Memory> findByMemoryType(MemoryType memoryType, Long lastMemoryId, int size);
    
    // 게임용 메서드들
    List<Memory> findMemoriesWithImagesByMember(Member member);
    List<Memory> findMemoriesWithImagesByMemoryType(MemoryType memoryType);

    Optional<Memory> findMemoryById(Long memoryId);
}

File: memory-domain/src/main/kotlin/com/memory/domain/memory/repository/MemoryRepositoryCustom.kt
package com.memory.domain.memory.repository

import com.memory.domain.memory.Memory
import com.memory.domain.member.Member
import com.memory.domain.memory.MemoryType
import java.util.Optional

interface MemoryRepositoryCustom {
    fun findByMemberAndMemoryType(member: Member?, relatedMemberIds: List<Long>?, memoryType: MemoryType?, size: Int): List<Memory>
    fun findByMemberAndMemoryType(member: Member?, relatedMemberIds: List<Long>?, memoryType: MemoryType?, lastMemoryId: Long?, size: Int): List<Memory>
    fun findMemoryByIdAndMemberId(memoryId: Long?, memberId: Long?): Optional<Memory>
    fun findByMemoryType(memoryType: MemoryType?, size: Int): List<Memory>
    fun findByMemoryType(memoryType: MemoryType?, lastMemoryId: Long?, size: Int): List<Memory>

    // 게임용 메서드들
    fun findMemoriesWithImagesByMember(member: Member?): List<Memory>
    fun findMemoriesWithImagesByMemoryType(memoryType: MemoryType?): List<Memory>
    fun findMemoryById(memoryId: Long?): Optional<Memory>
}
